---
title: "Read Homer Output Analysis"
output: html_notebook
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Cmd+Shift+Enter*. 

```{r setup, include=FALSE}
plot(cars)
```

Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Cmd+Option+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Cmd+Shift+K* to preview the HTML file). 

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.

---

Can you pull out a couple examples of TFs (maybe 1 or 2) from these 73K and identify which part of the motif is affected by the SNP? I bet it will be the part of a motif where there is a very heavy weight on one particular nucleotide and messing that up results in failure to identify the motif with the alternative allele. 

also, could you show one or two examples of TFs from the 69K set, where searching the reference genome fails to identify a motif but searching with the alternative allele finds it? 

Similarly, I bet the SNPs occur in the super important core part of the motif with high probabilities (from the PWM) for that nucleotide. Do you have experience making sequence logos? theyâ€™re fun plots to show the important of each nucleotide in a motif?

```{r load_libraries}
source("https://bioconductor.org/biocLite.R")
BiocManager::install("seqLogo")

library(seqLogo) # makePWM
library(stringr) # read.fasta, write.fasta
library(data.table)
library(dplyr)
library(glue)
```

```{r load_data}

in_ref_not_alt_unique_SNP <- fread("/Users/kathryntsai/OneDrive\ -\ Villanova\ University/College/2018-2019/Summer\ 2019/TFs_eQTLs_Research/RProjects/Project2/q2_3_output/in_ref_not_alt_unique_SNP.txt")
in_alt_not_ref_unique_SNP <- fread("/Users/kathryntsai/OneDrive\ -\ Villanova\ University/College/2018-2019/Summer\ 2019/TFs_eQTLs_Research/RProjects/Project2/q2_3_output/in_alt_not_ref_unique_SNP.txt")

x1 <- in_ref_not_alt_unique_SNP[!duplicated(in_ref_not_alt_unique_SNP),] # 73345
x2 <- in_alt_not_ref_unique_SNP[!duplicated(in_alt_not_ref_unique_SNP),] # 69674

a_gene_snp <- fread("/Users/kathryntsai/OneDrive\ -\ Villanova\ University/College/2018-2019/Summer\ 2019/TFs_eQTLs_Research/RProjects/Project2/q2_3_output/a_gene_snp.txt")

```

Using only motifs on positive strands, sort motifs by Motif Score
```{r get_samples}

x1_s <- x1[!duplicated(x1$"Motif Name"), ]
x2_s <- x2[!duplicated(x2$"Motif Name"), ]

z <- inner_join(x1_s[x1_s$"ref_Strand" == "+",], x2_s[x2_s$"alt_Strand" == "+",], by="Motif Name")
z_test_group <- tail(z[order(z$"ref_MotifScore", decreasing=F), c("Motif Name", "ref_MotifScore", "alt_MotifScore")], n=11L)

```

http://zevross.com/blog/2014/08/05/using-the-r-function-anti_join-to-find-unmatched-records/
```{r understanding_anti_join, include=F}
table1[!table1$id%in%table2$id,]
table1[is.na(match(table1$id,table2$id)),]
anti_join(table1, table2, by="id")
```

```{r display_data, include=F}

ref_test_group_transposed <- t(x1_s[x1_s$"Motif Name" %in% z_test_group$"Motif Name",c("SNPWindow", "ref_Offset", "Motif Name", "Sequence")])
alt_test_group_transposed <- t(x2_s[x2_s$"Motif Name" %in% z_test_group$"Motif Name",c("SNPWindow", "alt_Offset", "Motif Name", "Sequence")])

ref_test_group_transposed
alt_test_group_transposed
```

```{r test_group_data}
ref_test_group <- x1_s[x1_s$"Motif Name" %in% z_test_group$"Motif Name",c("SNPWindow", "ref_Offset", "Motif Name", "Sequence")]
alt_test_group <- x2_s[x2_s$"Motif Name" %in% z_test_group$"Motif Name",c("SNPWindow", "alt_Offset", "Motif Name", "Sequence")]

# Note: ref_test_group[1,"Motif Name"] != ref_test_group$`Motif Name`[1]
```


Motifs that will not adhere to pipeline: ap2gamma, batf3-irf8, cebp-cebp, cebp-ap1, cMyc.lncap, ctcf-mys, ets-distal, ets-ebox, ets-runx, ews-erg, ews-fli1, fox-ebox, gata3.dr4/dr8/ir3/ir4, gfy-staf, gre-raw, jun-ap1, jun-cre, nf1-fox, nf1-half, nfat-ap1, nkx2.1, nkx2.2, nkx2.5, oct-oct, oct-oct.short, oct.ir, oct.ir1, oct4-sox2, oct4-sox17, p53-myc, pax3-fkh4, pax5-short, pax7-long, pax7-longest,pu1-irf, pu1-irf8, rpbj1-ebox, stat3.il23, stat6.2, tbox-smad

```{r get_headers_from_homer}

if (read.table("/Users/kathryntsai/OneDrive\ -\ Villanova\ University/College/2018-2019/Summer\ 2019/TFs_eQTLs_Research/RProjects/Project2/q2_3_investigation/motifs/ronin.motif", sep="\t", header=F,nrow=1)[,2]

```


```{r get_motifs_from_homer}

pax7 <- read.table("/Users/kathryntsai/OneDrive\ -\ Villanova\ University/College/2018-2019/Summer\ 2019/TFs_eQTLs_Research/RProjects/Project2/q2_3_investigation/motifs/pax7-longest.motif", sep="\t", skip=1, header=F, stringsAsFactors=FALSE)
foxa1 <- read.table("motifs/foxa1.lncap.motif", skip=1, sep="\t", header=F, stringsAsFactors=FALSE)
ronin <- read.table("motifs/ronin.motif", skip=1, sep="\t", header=F, stringsAsFactors=FALSE)

```

```{r make_PWM}

pe <- makePWM(t(pax7))
pdf("Pax7_longest_bitsscore.pdf")
seqLogo(pe,ic.scale=T)
dev.off()

pe <- makePWM(t(foxa1))
pdf("Foxa1_lncap.bitsscore.pdf")
seqLogo(pe,ic.scale=T)
dev.off()
# AAAGTAAACA

pe <- makePWM(t(ronin))
pdf("Ronin_bitsscore.pdf")
seqLogo(pe,ic.scale=T)
dev.off()

```
```{r compare_TFs_and_sequence}

ref_pax7 <- (x1_s[x1_s$"Motif Name"=="Pax7(Paired,Homeobox),longest/Myoblast-Pax7-ChIP-Seq(GSE25064)/Homer", c("SNPWindow", "ref_Offset", "Motif Name", "ref_Strand", "Sequence")])
alt_pax7 <- (x2_s[x2_s$"Motif Name"=="Pax7(Paired,Homeobox),longest/Myoblast-Pax7-ChIP-Seq(GSE25064)/Homer", c("SNPWindow", "alt_Offset", "Motif Name", "alt_Strand", "Sequence")])

ref_ronin <- (x1_s[x1_s$"Motif Name"=="Ronin(THAP)/ES-Thap11-ChIP-Seq(GSE51522)/Homer", c("SNPWindow", "ref_Offset", "Motif Name", "ref_Strand", "Sequence")])
alt_ronin <- (x2_s[x2_s$"Motif Name"=="Ronin(THAP)/ES-Thap11-ChIP-Seq(GSE51522)/Homer", c("SNPWindow", "alt_Offset", "Motif Name", "alt_Strand", "Sequence")])

ref_foxa1 <- (x1_s[x1_s$"Motif Name"=="FOXA1:AR(Forkhead,NR)/LNCAP-AR-ChIP-Seq(GSE27824)/Homer", c("SNPWindow", "ref_Offset", "Motif Name", "ref_Strand", "Sequence")])
alt_foxa1 <- (x2_s[x2_s$"Motif Name"=="FOXA1:AR(Forkhead,NR)/LNCAP-AR-ChIP-Seq(GSE27824)/Homer", c("SNPWindow", "alt_Offset", "Motif Name", "alt_Strand", "Sequence")])

a_gene_snp$SNPWindow <- paste("chromosome:GRCh37:",a_gene_snp$SNPChr, ":", a_gene_snp$SNPWindowStart, ":", a_gene_snp$SNPWindowEnd,":1", sep="")

ref_pax7 <- merge(ref_pax7, a_gene_snp, by = c("SNPWindow"))
alt_pax7 <- merge(alt_pax7, a_gene_snp, by = c("SNPWindow"))
ref_ronin <- merge(ref_ronin, a_gene_snp, by = c("SNPWindow"))
alt_ronin <- merge(alt_ronin, a_gene_snp, by = c("SNPWindow"))
ref_foxa1 <- merge(ref_foxa1, a_gene_snp, by = c("SNPWindow"))
alt_foxa1 <- merge(alt_foxa1, a_gene_snp, by = c("SNPWindow"))
```

pax7_motif: xTAATTAxxTAATTAxxxxx
ref_sequence: AAAATTTTCAAATTGTAAAA, offset=-5
ref_pax7_fasta_seq_mod: TATAAAGTTTCTTCTAAAAATTTTCAAATTGTAAAATGTACA

alt_sequence: AAAATTAAAAAATTAAAATA, offset=-5
alt_pax7_fasta_seq_mod: TATAAAGTTTCTTCTAAAAGTTTTCAAATTGTAAAATGTACA <!--ACTCCATTTCAAAAACAAAATTAAAAAATTAAAATAAAAAAG-->

ronin_motif: GACTACAACTCCCAGAAGGC
ref_sequence: GACTCCAGCTCCCGGCAGGC, offset=-21
ref_ronin_fasta_seq_mod: GACTCCAGCTCCCGGCAGGCCTCGCGCGGCCCGGGGCCAGCG

alt_sequence: GACTACGAGTCCCACAATAA, offset=-8
alt_ronin_fasta_seq_mod: GACTCCAGCTCCCGGCAGGTCTCGCGCGGCCCGGGGCCAGCG

foxa1_motif: AAAGTAAACA
ref_sequence: AAAAAACAAAACAAAACAAA, offset=-21
ref_foxa1_fasta_seq_mod: AAAAAACAAAACAAAACAAAACAAAACAAAACAAAAAACTAA

alt_sequence: GGCAAACTTATAAGTACACA, offset=-8
alt_foxa1_fasta_seq_mod: AAAAAACAAAACAAAACAACACAAAACAAAACAAAAAACTAA


```{r}

```


