---
title: "Read Homer Output Analysis - For Ref/Alt"
output: html_notebook
---
---

Can you pull out a couple examples of TFs (maybe 1 or 2) from these 73K and identify which part of the motif is affected by the SNP? I bet it will be the part of a motif where there is a very heavy weight on one particular nucleotide and messing that up results in failure to identify the motif with the alternative allele. 

also, could you show one or two examples of TFs from the 69K set, where searching the reference genome fails to identify a motif but searching with the alternative allele finds it? 

Similarly, I bet the SNPs occur in the super important core part of the motif with high probabilities (from the PWM) for that nucleotide. Do you have experience making sequence logos? theyâ€™re fun plots to show the important of each nucleotide in a motif?

Limitations: only looks at positive strand

```{r load_libraries}
# source("https://bioconductor.org/biocLite.R")
# BiocManager::install("seqLogo")

library(seqLogo) # makePWM
library(stringr) # read.fasta, write.fasta
library(data.table)
library(dplyr)
library(glue)
library(matrixStats)
library(ggthemes)
```

```{r load_data}

in_ref_not_alt_unique_SNP <- fread("/Users/kathryntsai/OneDrive\ -\ Villanova\ University/College/2018-2019/Summer\ 2019/TFs_eQTLs_Research/RProjects/Project2/q2_3_output/in_ref_not_alt_unique_SNP.txt") # 264905 x 8
in_alt_not_ref_unique_SNP <- fread("/Users/kathryntsai/OneDrive\ -\ Villanova\ University/College/2018-2019/Summer\ 2019/TFs_eQTLs_Research/RProjects/Project2/q2_3_output/in_alt_not_ref_unique_SNP.txt") # 269853 x 8

x1 <- in_ref_not_alt_unique_SNP[!duplicated(in_ref_not_alt_unique_SNP),] # 70550 x 8
x2 <- in_alt_not_ref_unique_SNP[!duplicated(in_alt_not_ref_unique_SNP),] # 72078 x 8
#in_ref_not_alt_unique_SNP$ref_Offset <- in_ref_not_alt_unique_SNP$ref_Offset + 21
#in_alt_not_ref_unique_SNP$alt_Offset <- in_alt_not_ref_unique_SNP$alt_Offset + 21

a_gene_snp <- fread("/Users/kathryntsai/OneDrive\ -\ Villanova\ University/College/2018-2019/Summer\ 2019/TFs_eQTLs_Research/RProjects/Project2/q2_3_output/a_gene_snp.txt") # 925540 x 14
a_gene_snp$SNPWindow <- paste("chromosome:GRCh37:",a_gene_snp$SNPChr, ":", a_gene_snp$SNPWindowStart, ":", a_gene_snp$SNPWindowEnd,":1", sep="") # 925540 x 15
```

Using only motifs on positive strands, sort motifs by Motif Score
Change # data points with value of n in z_test_group
```{r get_samples}

# x1_s <- x1 %>% sample_n(800)
# x2_s <- x2 %>% sample_n(800)
# fwrite(x1_s, "q2_3_output/x1_s_ref.txt", sep="\t")
# fwrite(x2_s, "q2_3_output/x2_s_alt.txt", sep="\t")

x1_s <- fread("q2_3_output/x1_s_ref.txt", sep='\t')
x2_s <- fread("q2_3_output/x2_s_alt.txt", sep='\t')

z <- x1_s[x1_s$"ref_Strand" == "+",]
ref_test_group <- z[order(z$"ref_MotifScore", decreasing=F),]

z <- x2_s[x2_s$"alt_Strand" == "+",]
alt_test_group <- z[order(z$"alt_MotifScore", decreasing=F),]

```


Todo: detect # motifs without hardcoding it in

```{r create_motifs_table}

# used forklift to rename all the motifs

motifs_table <- data.frame(read.table("/Users/kathryntsai/OneDrive\ -\ Villanova\ University/College/2018-2019/Summer\ 2019/TFs_eQTLs_Research/RProjects/Project2/q2_3_investigation/motifs/motif1.motif", sep="\t", header=F,nrow=1)[,2])
colnames(motifs_table) <- "Motif Name"

n = 429
for (i in 2:n){
  z <- read.table(paste("/Users/kathryntsai/OneDrive\ -\ Villanova\ University/College/2018-2019/Summer\ 2019/TFs_eQTLs_Research/RProjects/Project2/q2_3_investigation/motifs/motif", i, ".motif", sep=""), sep="\t", header=F,nrow=1)[,2]
  z <- data.frame(z)
  colnames(z) <- "Motif Name"
  motifs_table <- rbind(motifs_table, z)
}

motifs_table <- cbind(motifs_table, 1:n)
colnames(motifs_table)[2] <- "Num"
# 429 x 2

```

```{r match_motifs_with_test_groups_ref}
 
ref_test_group_file <- left_join(ref_test_group, motifs_table) # 10051 x 9
ref_test_group_file$SNPPosInFasta <- 0.001 

ref_test_group_with_fasta <- left_join(ref_test_group_file, a_gene_snp[!duplicated(a_gene_snp$fasta_seq_og),]) # also by 'SNP'
ref_test_group_with_fasta$ref_Offset <- ref_test_group_with_fasta$ref_Offset + 21
ref_test_group_with_fasta <- ref_test_group_with_fasta[which(str_sub(ref_test_group_with_fasta$fasta_seq_og, 21, 21) == ref_test_group_with_fasta$OtherAllele), ] # only necessary if you didn't use old a_gene_snp file for homer input
counter_of_excluded_data <- 0

```

```{r get_max_pwm_info_ref}

i=1

## make_PWM_and_get_motif_sequence_for_1_motif
impt_table <- read.table(paste("/Users/kathryntsai/OneDrive\ -\ Villanova\ University/College/2018-2019/Summer\ 2019/TFs_eQTLs_Research/RProjects/Project2/q2_3_investigation/motifs/motif", ref_test_group_with_fasta[i,"Num"], ".motif", sep=""), sep="\t", skip=1, header=F, stringsAsFactors=FALSE)
colnames(impt_table) <- c("A", "C", "G", "T")
max_TF_motifs <- data.frame(cbind(data.frame(rowMaxs(as.matrix(impt_table), cols=T)), 1:nrow(impt_table), colnames(impt_table)[apply(impt_table,1,which.max)]))
colnames(max_TF_motifs) <- c("Num", "Index", "Base")
tf_motif <- paste(max_TF_motifs$Base, collapse="", sep="")

if (ref_test_group_with_fasta$ref_Offset[i] >= (21 - nchar(tf_motif)) & ref_test_group_with_fasta$ref_Offset[i] < 21)
{
    ref_test_group_with_fasta$SNPPosInFasta[i] <- 21 - ref_test_group_with_fasta$ref_Offset[i]+1
    all_snp_pwms_ref <- max_TF_motifs[21 - ref_test_group_with_fasta$ref_Offset[i], ] # make this rbind later
    all_snp_pwms_ref <- cbind(all_snp_pwms_ref, i, ref_test_group_with_fasta[i,]) # make this cbind later
} else {
  all_snp_pwms_ref <- data.table(NA, NA, NA, i, ref_test_group_with_fasta[i]) # to begin
  colnames(all_snp_pwms_ref) <- c("Num", "Index", "Base", "i", colnames(ref_test_group_with_fasta))
  counter_of_excluded_data <- counter_of_excluded_data + 1
}
    
for (i in 2:nrow(ref_test_group_with_fasta)){
  impt_table <- read.table(paste("/Users/kathryntsai/OneDrive\ -\ Villanova\ University/College/2018-2019/Summer\ 2019/TFs_eQTLs_Research/RProjects/Project2/q2_3_investigation/motifs/motif", ref_test_group_with_fasta[i,"Num"], ".motif", sep=""), sep="\t", skip=1, header=F, stringsAsFactors=FALSE)
  colnames(impt_table) <- c("A", "C", "G", "T")
  max_TF_motifs <- data.frame(cbind(data.frame(rowMaxs(as.matrix(impt_table), cols=T)), 1:nrow(impt_table), colnames(impt_table)[apply(impt_table,1,which.max)]))
  colnames(max_TF_motifs) <- c("Num", "Index", "Base")
  tf_motif <- paste(max_TF_motifs$Base, collapse="", sep="")  # CATGCATATGCATAT
  
  ## does_TF_hit_SNP_if_so_get_position
  if (ref_test_group_with_fasta$ref_Offset[i] >= (21 - nchar(tf_motif)) & ref_test_group_with_fasta$ref_Offset[i] < 21) # if offset = 21, skip this
  {
      ref_test_group_with_fasta$SNPPosInFasta[i] <- 21 - ref_test_group_with_fasta$ref_Offset[i]
      row_to_bind <- data.frame(max_TF_motifs[21 - ref_test_group_with_fasta$ref_Offset[i], ], i, ref_test_group_with_fasta[i,])
      colnames(row_to_bind) <- colnames(all_snp_pwms_ref)
      all_snp_pwms_ref <- rbind(all_snp_pwms_ref, row_to_bind)
  } else {
    counter_of_excluded_data <- counter_of_excluded_data + 1
  }
}

all_snp_pwms_ref <- all_snp_pwms_ref[!is.na(all_snp_pwms_ref$Num),]

```

```{r visualize_likelihood_ref}

colnames(all_snp_pwms_ref) <- c("Num", "Index", "Base", "i", "SNPWindow", "Offset", "Sequence", "Motif Name", "Strand", "MotifScore", "Num1", "SNPPosInFasta", "Gene", "GeneSymbol", "GeneChr", "GenePos", "SNP", "SNPChr", "SNPPos", "AssessedAllele", "OtherAllele", "SNPWindowStart", "SNPWindowEnd", "fasta_seq_og", "fasta_seq_mod")

ggplot(all_snp_pwms_ref, aes(x=as.numeric(Num))) + 
  geom_histogram(binwidth=.1, color="black", fill="blue") +
  theme_stata() +
  geom_density(alpha=.2, fill="#FF6666", aes(y=..count../8)) +
  geom_vline(aes(xintercept=mean(as.numeric(Num))),
            color="blue", linetype="dashed", size=1) + 
  ggtitle("Likelihood of All Motifs - Reference Allele")

```

For alt

```{r match_motifs_with_test_groups_alt}

alt_test_group_file <- left_join(alt_test_group, motifs_table) # 9901 x 9
alt_test_group_file$SNPPosInFasta <- 0.001 

alt_test_group_with_fasta <- left_join(alt_test_group_file, a_gene_snp[!duplicated(a_gene_snp$fasta_seq_og),]) # also by 'SNP'
alt_test_group_with_fasta$alt_Offset <- alt_test_group_with_fasta$alt_Offset + 21
alt_test_group_with_fasta <- alt_test_group_with_fasta[which(str_sub(alt_test_group_with_fasta$fasta_seq_og, 21, 21) == alt_test_group_with_fasta$OtherAllele), ] # 37x20, only necessary if you didn't use old a_gene_snp file for homer input
counter_of_excluded_data <- 0

```

```{r get_max_pwm_info_alt}

i=1

## make_PWM_and_get_motif_sequence_for_1_motif
impt_table <- read.table(paste("/Users/kathryntsai/OneDrive\ -\ Villanova\ University/College/2018-2019/Summer\ 2019/TFs_eQTLs_Research/RProjects/Project2/q2_3_investigation/motifs/motif", alt_test_group_with_fasta[i,"Num"], ".motif", sep=""), sep="\t", skip=1, header=F, stringsAsFactors=FALSE)
colnames(impt_table) <- c("A", "C", "G", "T")
max_TF_motifs <- data.frame(cbind(data.frame(rowMaxs(as.matrix(impt_table), cols=T)), 1:nrow(impt_table), colnames(impt_table)[apply(impt_table,1,which.max)]))
colnames(max_TF_motifs) <- c("Num", "Index", "Base")
tf_motif <- paste(max_TF_motifs$Base, collapse="", sep="")

## get_fasta_seq_and_get_rid_of_matches_that_do_not_work: Where does the SNP hit the TF? for alt_test_group: I think eQTLGen is zero-indexed.

## does_TF_hit_SNP_if_so_get_position
if (alt_test_group_with_fasta$alt_Offset[i] >= (21 - nchar(tf_motif)) & alt_test_group_with_fasta$alt_Offset[i] < 21)
{
    alt_test_group_with_fasta$SNPPosInFasta[i] <- 21 - alt_test_group_with_fasta$alt_Offset[i]+1
    all_snp_pwms_alt <- max_TF_motifs[21 - alt_test_group_with_fasta$alt_Offset[i], ] # make this rbind later
    all_snp_pwms_alt <- cbind(all_snp_pwms_alt, i, alt_test_group_with_fasta[i,]) # make this cbind later
} else {
  all_snp_pwms_alt <- data.table(NA, NA, NA, i, alt_test_group_with_fasta[i,]) # to begin
  colnames(all_snp_pwms_alt) <- c("Num", "Index", "Base", "i", colnames(alt_test_group_with_fasta))
  counter_of_excluded_data <- counter_of_excluded_data + 1
}
    
for (i in 2:nrow(alt_test_group_with_fasta)){
  impt_table <- read.table(paste("/Users/kathryntsai/OneDrive\ -\ Villanova\ University/College/2018-2019/Summer\ 2019/TFs_eQTLs_Research/RProjects/Project2/q2_3_investigation/motifs/motif", alt_test_group_with_fasta[i,"Num"], ".motif", sep=""), sep="\t", skip=1, header=F, stringsAsFactors=FALSE)
  colnames(impt_table) <- c("A", "C", "G", "T")
  max_TF_motifs <- data.frame(cbind(data.frame(rowMaxs(as.matrix(impt_table), cols=T)), 1:nrow(impt_table), colnames(impt_table)[apply(impt_table,1,which.max)]))
  colnames(max_TF_motifs) <- c("Num", "Index", "Base")
  tf_motif <- paste(max_TF_motifs$Base, collapse="", sep="")  # CATGCATATGCATAT
  
  ## does_TF_hit_SNP_if_so_get_position
  if (alt_test_group_with_fasta$alt_Offset[i] >= (21 - nchar(tf_motif)) & alt_test_group_with_fasta$alt_Offset[i] < 21) # if offset = 21, skip this
  {
      alt_test_group_with_fasta$SNPPosInFasta[i] <- 21 - alt_test_group_with_fasta$alt_Offset[i]
      row_to_bind <- data.frame(max_TF_motifs[21 - alt_test_group_with_fasta$alt_Offset[i], ], i, alt_test_group_with_fasta[i,])
      colnames(row_to_bind) <- colnames(all_snp_pwms_alt)
      all_snp_pwms_alt <- rbind(all_snp_pwms_alt, row_to_bind)
  } else {
    counter_of_excluded_data <- counter_of_excluded_data + 1
  }
}

all_snp_pwms_alt <- all_snp_pwms_alt[!is.na(all_snp_pwms_alt$Num),]
```

```{r visualize_likelihood_alt}

colnames(all_snp_pwms_alt) <- c("Num", "Index", "Base", "i", "SNPWindow", "Offset", "Sequence", "Motif Name", "Strand", "MotifScore", "Num1", "SNPPosInFasta", "Gene", "GeneSymbol", "GeneChr", "GenePos", "SNP", "SNPChr", "SNPPos", "AssessedAllele", "OtherAllele", "SNPWindowStart", "SNPWindowEnd", "fasta_seq_og", "fasta_seq_mod")

ggplot(all_snp_pwms_alt, aes(x=as.numeric(Num))) + 
  geom_histogram(binwidth=.1, color="black", fill="blue") +
  theme_stata() +
  geom_density(alpha=.2, fill="#FF6666", aes(y=..count../8)) +
  geom_vline(aes(xintercept=mean(as.numeric(Num))),
            color="blue", linetype="dashed", size=1) + 
  ggtitle("Likelihood of All Motifs - Alternative Allele")

```
